glass/grf/_transformations.py

```diff
@@ -0,0 +1,178 @@
+        """Return *x* unchanged."""
+        return x
+
+
+@dataclass
+class Lognormal:
+    r"""
+    Transformation for lognormal fields.
+
+    .. math::
+
+       t(X) = \lambda \bigl[\exp(X - \tfrac{\sigma^2}{2}) - 1\bigr]
+
+    The "lognormal shift" parameter :math:`\lambda` is the scale
+    parameter of the distribution, with determines the smallest
+    possible value :math:`-\lambda` of the lognormal field.
+
+    """
+
+    lamda: float = 1.0
+
+    def __call__(self, x: NDArray[Any], var: float, /) -> NDArray[Any]:
+        """Transform *x* into a lognormal field."""
+        xp = x.__array_namespace__()
+        x = xp.expm1(x - var / 2)
+        if self.lamda != 1.0:
+            x = self.lamda * x
+        return x
+
+
+@dataclass
+class SquaredNormal:
+    r"""
+    Transformation for squared normal fields as introduced by
+    [Tessore25]_.
+
+    .. math::
+
+       t(X) = \lambda \, \bigl[(X - a)^2 - 1\bigr]
+
+    The transformation is characterised by the shape parameter *a*,
+    which is related to the variance :math:`\sigma^2` of the Gaussian
+    random field, :math:`a = \sqrt{1 - \sigma^2}`.
+
+    The "shift" parameter :math:`\lambda` is the scale parameter of the
+    distribution, with determines the smallest possible value
+    :math:`-\lambda` of the squared normal field.
+
+    .. [Tessore25] https://arxiv.org/abs/2408.16903
+
+    """  # noqa: D205
+
+    a: float
+    lamda: float = 1.0
+
+    def __call__(self, x: NDArray[Any], _var: float, /) -> NDArray[Any]:
+        """Transform *x* into a squared normal field."""
+        x = (x - self.a) ** 2 - 1
+        if self.lamda != 1.0:
+            x = self.lamda * x
+        return x
+
+
+########################################################################
+# normal x normal
+########################################################################
+
+
+@corr.add
+def _(_t1: Normal, _t2: Normal, x: NDArray[Any], /) -> NDArray[Any]:
+    return x
+
+
+@icorr.add
+def _(_t1: Normal, _t2: Normal, x: NDArray[Any], /) -> NDArray[Any]:
+    return x
+
+
+@dcorr.add
+def _(_t1: Normal, _t2: Normal, x: NDArray[Any], /) -> NDArray[Any]:
+    return 1.0 + (0 * x)
+
+
+########################################################################
+# lognormal x lognormal
+########################################################################
+
+
+@corr.add
+def _(t1: Lognormal, t2: Lognormal, x: NDArray[Any], /) -> NDArray[Any]:
+    xp = x.__array_namespace__()
+    return t1.lamda * t2.lamda * xp.expm1(x)  # type: ignore[no-any-return]
+
+
+@icorr.add
+def _(t1: Lognormal, t2: Lognormal, x: NDArray[Any], /) -> NDArray[Any]:
+    xp = x.__array_namespace__()
+    return xp.log1p(x / (t1.lamda * t2.lamda))  # type: ignore[no-any-return]
+
+
+@dcorr.add
+def _(t1: Lognormal, t2: Lognormal, x: NDArray[Any], /) -> NDArray[Any]:
+    xp = x.__array_namespace__()
+    return t1.lamda * t2.lamda * xp.exp(x)  # type: ignore[no-any-return]
```

Comment:
This PR is introducing a lot of things that are breaking type-checking. I assume you've tried to fix them?

Reply:
This isn't really breaking type checking. The stubs don't currently know `xp.exp()` is returning an array.

Reply:
I had my own Array and ArrayNamespace stubs here to make the type-checking work fully, btw, so I'm also happy to put these back until data-apis/array-api-typing becomes functional.